src/storage/write_ahead_log/log_serializer_task.cpp

```diff
@@ -148,26 +184,44 @@ void LogSerializerTask::HandFilledBufferToWriter() {
   NOISEPAGE_ASSERT(primary_replication_manager_ != DISABLED ||
                        retention_policy != transaction::RetentionPolicy::RETENTION_LOCAL_DISK_AND_NETWORK_REPLICAS,
                    "If replication is disabled, then you can't send buffers to replicas.");
+  if (destination == SerializeDestination::DISK) {
+    if (disk_filled_buffer_ != nullptr) {
+      disk_filled_buffer_->PrepareForSerialization(retention_policy);
+    }
+    // Hand over the filled buffer
+    disk_filled_buffer_queue_->Enqueue(std::make_pair(disk_filled_buffer_, disk_commits_in_buffer_));
+    // Signal disk log consumer task thread that a buffer has been handed over
+    disk_log_writer_thread_cv_->notify_one();
+    // Mark that the task doesn't have a buffer in its possession to which it can write to
+    disk_commits_in_buffer_.clear();
+    disk_filled_buffer_ = nullptr;
+  } else if (destination == SerializeDestination::REPLICAS) {
+    if (replication_filled_buffer_ != nullptr) {
+      replication_filled_buffer_->PrepareForSerialization(retention_policy);
+
+      // Replicate the buffer if it exists.
+      // TODO(WAN): Note that this is effectively on the critical path to commit callbacks being invoked. Aka terrible.
+      if (primary_replication_manager_ != DISABLED) {
+        primary_replication_manager_->ReplicateBuffer(replication_filled_buffer_);
+      }
+    }
 
-  if (filled_buffer_ != nullptr) {
-    filled_buffer_->PrepareForSerialization(retention_policy);
-    // Replicate the buffer if it exists.
-    // TODO(WAN): Note that this is effectively on the critical path to commit callbacks being invoked. Aka terrible.
-    if (primary_replication_manager_ != DISABLED) {
-      primary_replication_manager_->ReplicateBuffer(filled_buffer_);
+    // Hand over the filled buffer
+    if (replication_filled_buffer_ != nullptr) {
```

Comment:
This is definitely echoing @mbutrovich comment, but you can merge this whole block with the previous one which has the same check in the if statement.